### ISSUE TITLE
Build failure

### ISSUE THREAD

**Issue by IzzySoft** [leohearts/AlternativeUnlockXposed#6]
I've checked your app if its build is reproducible (see: [Reproducible bulds, special client support and more in our repo](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients?lang=en)), but I wasn't even able to compile from the tag as it seems some resources have not been checked in:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processReleaseResources'.
> A failure occurred while executing com.android.build.gradle.internal.res.LinkApplicationAndroidResourcesTask$TaskAction
   > Android resource linking failed
     com.leohearts.alternativeUnlockHook.app-main-66:/mipmap-anydpi-v26/ic_launcher.xml:3: error: resource color/ic_launcher_background (aka com.leohearts.alternativeUnlockHook:color/ic_launcher_background) not found.
     com.leohearts.alternativeUnlockHook.app-main-66:/mipmap-anydpi-v26/ic_launcher.xml:4: error: resource mipmap/ic_launcher_foreground (aka com.leohearts.alternativeUnlockHook:mipmap/ic_launcher_foreground) not found.
     com.leohearts.alternativeUnlockHook.app-main-66:/mipmap-anydpi-v26/ic_launcher_round.xml:3: error: resource color/ic_launcher_background (aka com.leohearts.alternativeUnlockHook:color/ic_launcher_background) not found.
     com.leohearts.alternativeUnlockHook.app-main-66:/mipmap-anydpi-v26/ic_launcher_round.xml:4: error: resource mipmap/ic_launcher_foreground (aka com.leohearts.alternativeUnlockHook:mipmap/ic_launcher_foreground) not found.
     error: failed linking file resources.
```

I've checked with the "nearby commits" the APK might have been built from, given its timestamp – but no dice, they stay AWOL. Could you please ensure all resources needed to build the app are properly checked in and pushed? Thanks a lot!

We'd appreciate if you could help making your build reproducible. We've prepared some [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) for that.

Looking forward to your reply!

**Comment by leohearts** [comment]
I'm digging into this problem.

**Comment by leohearts** [comment]
Should be fixed now 😶

**Comment by IzzySoft** [comment]
Thanks, that was fast! Will there be a new release soon, so I can retry for Reproducible Build? That needs a release with an APK attached that was built from the commit the tag points to. I see the last commit bumps the version, and there's a new tag already, so maybe…? Ah, looks like you're still on it, I just missed the "reopen" :see_no_evil: OK, waiting for the next ping then :crossed_fingers:

**Comment by leohearts** [comment]
![image](https://github.com/user-attachments/assets/554a2421-2e3f-4a95-ad85-77ce813edc4a)

Looks good to me

**Comment by IzzySoft** [comment]
Thanks! I confirm it builds now, but… I guess you first built and then committed?

```diff
-------------------------------
--- /dev/fd/63  2024-08-04 02:16:43.927873681 +0200
+++ /dev/fd/62  2024-08-04 02:16:43.927873681 +0200
@@ -1,7 +1,7 @@
   META-INF/com/android/build/gradle/app-metadata.properties
   32-bit CRC value (hex):                         965cb2d0
   META-INF/version-control-info.textproto
-  32-bit CRC value (hex):                         9805246e
+  32-bit CRC value (hex):                         eb37f163
   assets/dexopt/baseline.prof
   32-bit CRC value (hex):                         48dc1789
   assets/dexopt/baseline.profm
```

The embedded commit hash in that file says `b91e3312c0f9052bf8f88c77b9509962af0bdb90`, the tag points to `2b340b8dae2feb9469e6f01f55989f07b71b5f50` :wink: OK, let me see if a soft-reset makes it reproducible (that would trick the build to embed that commit hash without changing any files)… Yupp, congrats! That did it! So welcome to the community of Reproducible Builds then! If you remember §1 of the [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) with future releases, they should also be RB without any manual intervention needed:

> §1: Always build the APK from exactly the commit your release-tag will point to!

:wink:

Btw: your app is [listed with IzzyOnDroid](https://apt.izzysoft.de/packages/com.leohearts.alternativeUnlockHook) (with the next sync around 6 pm UTC it will also carry the green RB shield then). So be welcome to [pick a badge](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) to link there e.g. from your Readme!

PS: announcement here: https://floss.social/@IzzyOnDroid/112900944279515192

Thanks a lot for your help!

**Comment by leohearts** [comment]
> Thanks! I confirm it builds now, but… I guess you first built and then committed?
> 
> ```diff
> -------------------------------
> --- /dev/fd/63  2024-08-04 02:16:43.927873681 +0200
> +++ /dev/fd/62  2024-08-04 02:16:43.927873681 +0200
> @@ -1,7 +1,7 @@
>    META-INF/com/android/build/gradle/app-metadata.properties
>    32-bit CRC value (hex):                         965cb2d0
>    META-INF/version-control-info.textproto
> -  32-bit CRC value (hex):                         9805246e
> +  32-bit CRC value (hex):                         eb37f163
>    assets/dexopt/baseline.prof
>    32-bit CRC value (hex):                         48dc1789
>    assets/dexopt/baseline.profm
> ```
> 
> The embedded commit hash in that file says `b91e3312c0f9052bf8f88c77b9509962af0bdb90`, the tag points to `2b340b8dae2feb9469e6f01f55989f07b71b5f50` 😉 OK, let me see if a soft-reset makes it reproducible (that would trick the build to embed that commit hash without changing any files)… Yupp, congrats! That did it! So welcome to the community of Reproducible Builds then! If you remember §1 of the [hints on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds) with future releases, they should also be RB without any manual intervention needed:
> 
> > §1: Always build the APK from exactly the commit your release-tag will point to!
> 
> 😉
> 
> Btw: your app is [listed with IzzyOnDroid](https://apt.izzysoft.de/packages/com.leohearts.alternativeUnlockHook) (with the next sync around 6 pm UTC it will also carry the green RB shield then). So be welcome to [pick a badge](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) to link there e.g. from your Readme!
> 
> PS: announcement here: https://floss.social/@IzzyOnDroid/112900944279515192
> 
> Thanks a lot for your help!

Wow that's a lesson to me, I never knew apk metadata has git commit hash before . And thanks for your build and repo maintenance ♥️ I'm a user of lzzyOnDroid too.

**Comment by IzzySoft** [comment]
> I never knew apk metadata has vc hash before

That's also something relatively new, introduced with some AGP update (don't ask me which one, I don't remember – but IIRC it was within the last 12 months). While I initially thought it a nuisance, I meanwhile figured this one is quite useful. In your case here maybe not – but sometimes devs attach an APK built from a different commit and switching to that (not via soft-reset but via checkout) helps achieving RB. So this information tells me which commit to try. And of course to tell the dev about "§1" – I guess about 4 out of 5 of RB failures are caused by "violating" that.

> I'll try to ensure every build to be from the exact tag for future builds.

:star_struck: I'll adjust [your recipe](https://codeberg.org/IzzyOnDroid/rbtlog/src/branch/izzy/log/recipes/com.leohearts.alternativeUnlockHook.yml) with the next release then, so it no longer requires manual intervention from that one onwards.

> And thanks for your build and repo maintenance ♥️

Gladly! Thanks for your warm feedback! That's a reward in itself. I'm happy enough to say that feedback is positive almost always, which of course gives me motivation to carry on. I'm also glad having a team now; especially with RBs, I hardly could do all this alone anymore…

> I'm a user of lzzyOnDroid too.

Yay, nice! Should you also use Magisk (pretty likely if you use Xposed), did you know I run a [magisk repo](https://apt.izzysoft.de/magisk) as well?

**Comment by leohearts** [comment]
>  did you know I run a magisk repo as well?
No but I know it now. Thank you !

**Comment by IzzySoft** [comment]
Latest release (1.2v2) again fails, this time with:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:dataBindingMergeDependencyArtifactsRelease'.
> Could not resolve all files for configuration ':app:releaseCompileClasspath'.
   > Could not find de.robv.android.xposed:api:82.
     Searched in the following locations:
       - https://dl.google.com/dl/android/maven2/de/robv/android/xposed/api/82/api-82.pom
       - https://repo.maven.apache.org/maven2/de/robv/android/xposed/api/82/api-82.pom
       - https://jcenter.bintray.com/de/robv/android/xposed/api/82/api-82.pom
     Required by:
         project :app
```

You might have that cached locally, but if I understand correctly (saw the same error with other modules as well), 82 is no longer available publicly (not sure, but maybe was on jcenter which finally [shut down](https://blog.gradle.org/jcenter-shutdown) a few days ago?

**Comment by leohearts** [comment]
Is the [official Xposed API repository](https://api.xposed.info/) allowed on IzzyOnDroid build server? F-droid [doesn't](https://gitlab.com/fdroid/fdroidserver/-/blob/df97ebe05298b6c0261d2dfbc1c2dd05687cfc7c/fdroidserver/scanner.py#L456-L479).

ref: https://github.com/deltazefiro/Amarok-Hider/issues/160

If we can't , I'll try to compile Xposed from source. (But should we really ?)

**Comment by IzzySoft** [comment]
We have no blacklist for Maven repos (at least not yet) – so you could give that a try, yes. Might be more transparent than just adding the JAR here directly. If that works out, it might also be interesting for https://github.com/dttzyjw0012/PhantomProcessesSavior/issues/4

> If we can't , I'll try to compile Xposed from source. (But should we really ?)

Well, building from source is a good thing. Look at the topic of this issue :stuck_out_tongue_winking_eye:

**Comment by leohearts** [comment]
> We have no blacklist for Maven repos (at least not yet) – so you could give that a try, yes. Might be more transparent than just adding the JAR here directly. If that works out, it might also be interesting for [dttzyjw0012/PhantomProcessesSavior#4](https://github.com/dttzyjw0012/PhantomProcessesSavior/issues/4)
> 
> > If we can't , I'll try to compile Xposed from source. (But should we really ?)
> 
> Well, building from source is a good thing. Look at the topic of this issue 😜

Should be fixed in v3 😶

**Comment by IzzySoft** [comment]
Please make sure to not forget increasing `versionCode` :wink: From the 24th:

    Binary files repo/com.leohearts.alternativeUnlockHook_3.apk and repo/com.leohearts.alternativeUnlockHook_1.2v2.apk differ

`versionCode` is what Android uses internally to keep versions apart, so this new release won't reach those having the 1.2 installed already. And it overwrote the APK from 1.2, as `<package_name>_<versionCode>` is a "unique key" there…

Oops: tag name is `1.2v2`, but release is named `1.2v3`, this is quite confusing. Did you somehow reuse the tag? Does it point to the commit the APK was built from? Else RB will fail.

**Comment by leohearts** [comment]
> Oops: tag name is `1.2v2`, but release is named `1.2v3`, this is quite confusing. Did you somehow reuse the tag? Does it point to the commit the APK was built from? Else RB will fail.

Oops, I forgot to change the tag with release. 

>Please make sure to not forget increasing versionCode 😉 From the 24th:
> so this new release won't reach those having the 1.2 installed already.

This is intended as it shouldn't affect those already using 1.2 .

**Comment by IzzySoft** [comment]
Well, that breaks several things then. Other platforms wouldn't even pull the update if the `versionCode` has not changed, so it would never be rolled out at all (that was the default here as well originally, and we might switch back to that at some point). And in combination with the tag name being the same, RB would never be confirmed as the APK no longer matches (and wouldn't be pulled for re-evaluation either as the tag was already processed).

Just saying to make implications known (and to avoid "unexpected side-effects"), not complaining here :wink:

**Comment by leohearts** [comment]
> Well, that breaks several things then. Other platforms wouldn't even pull the update if the `versionCode` has not changed, so it would never be rolled out at all (that was the default here as well originally, and we might switch back to that at some point). And in combination with the tag name being the same, RB would never be confirmed as the APK no longer matches (and wouldn't be pulled for re-evaluation either as the tag was already processed).
> 
> Just saying to make implications known (and to avoid "unexpected side-effects"), not complaining here :wink: 

Omg I could know it now 😶 I'll try to make another release after busy those days ♥️

**Comment by IzzySoft** [comment]
No worries – such implications are unclear to many people (which is why I throw in a few explaining words "just in case"). Just keep it in mind for the next releases. v3 was already pulled by the RB verificator here now (and by the repo itself yesterday it seems).

TL;DR: reusing code is fine, reusing `versionCode` not :see_no_evil:

And thanks a lot for your help there! :star_struck:

**Comment by IzzySoft** [comment]
PS: build run just finished here, 1.2v3 built fine and RB without any adjustments needed, :partying_face: Congrats & Thanks!

**Comment by IzzySoft** [comment]
Btw, before closing: you might wish to remove [this line](https://github.com/leohearts/AlternativeUnlockXposed/blob/master/settings.gradle.kts#L13) from your `settings.gradle.kts` – JCenter is dead, so that line just throws a "deprecation warning" currently without giving any benefit, @leohearts.

**Comment by leohearts** [comment]
Thanks for mention it but tbh I left it there on purpose. Removing it may also be a good option and I'll consider it next time on development 🩵

> Btw, before closing: you might wish to remove [this line](https://github.com/leohearts/AlternativeUnlockXposed/blob/master/settings.gradle.kts#L13) from your `settings.gradle.kts` – JCenter is dead, so that line just throws a "deprecation warning" currently without giving any benefit, @leohearts.

**Comment by IzzySoft** [comment]
No prob, as you see fit. I just noticed it in the build logs from my verification builder and thought to let you know :stuck_out_tongue_winking_eye:
